mem-ruby, configs: Add a CHI-TLM controller + testing

giactra opened the pull request:
This PR is implement a CacheController which acts as a bridge between the
AMBA TLM 2.0 implementation of CHI [[1]](https://developer.arm.com/documentation/101459/latest) [[2](https://developer.arm.com/Architectures/AMBA#Downloads)] with the gem5 (ruby) one.

In other words it translates AMBA CHI transactions into ruby
messages (which are then forwarded to the MessageQueues)
and viceversa.

```
ARM::CHI::Payload,         CHIRequestMsg
                     <-->  CHIDataMsg
ARM::CHI::Phase            CHIResponseMsg
                           CHIDataMsg
```

To be able to do so, it first adds a CHIGenericController class to model a coherence controller entirely
in C++ to work with ruby. Adding a CHIGenericController removes the limitation of using the gem5 Sequencer
and associated data structures, therefore allowing the aforementioned CHI to CHI bridge/controller
and potentially a hybrid setup with classic caches + ruby.

The PR also adds a CHI tester, which is a transaction generator directly connected to the CHI-TLM controller. It will
be possible to write small unit tests in python (the library has been wrapped with pybind) for CHI transactions and unlock unit-testing for ruby

tiagormk committed:
mem-ruby: sequencer prints panic pkts

Change-Id: I9cd780597c4680513d9cbeb8dda2e13f2a1faf56

tiagormk committed:
mem-ruby: Implement a CHI generic controller

Component implementing a generic controller that allow classic caches
interaction with Ruby/CHI.
The CHIGenericController provides an interface to send/receive CHI
messages to/from the interconnect. This is implement in C++ rather then
SLICC. This controller is seen as a MachineType:Cache by the CHI
implementation in SLICC.

Change-Id: I3afc4363f4290095c2f7428c8487bccd932e0300

tiagormk committed:
mem-ruby: add CHI SnpRespData_SD_Fwded_SC message

This snoop reponse is not generated internally by the SLICC
implementation, but is required for compatibility with classic caches
which may remain in SD state while returning SC data upon receiving
a converted SnpShared.

Change-Id: I5270b29c8863c7afd8abc39b3c7978b95330c183

tiagormk committed:
mem-ruby: add CHI missing valid SnpRespData type

Change-Id: I49c24e8b99932f8ae88511bb7a08a94f59ce7d29

giactra committed:
configs: Instantiate RNFs and MN via callbacks

This commit allows top level configs making use of the Ruby module
to define node generation callbacks.
The config_ruby function will check the system object for two
factory methods

1) _rnf_gen, if defined, will be called to generate RNFs
2) _mn_gen, if defined, will be called to generate MNs

Change-Id: I9daeece646e7cdb2d3bfefa761a9650562f8eb4b
Signed-off-by: Giacomo Travaglini <giacomo.travaglini@arm.com>

powerjg commented:
Would you be willing to talk about this in the next gem5-dev meeting?

This is super cool!

giactra commented:
> Would you be willing to talk about this in the next gem5-dev meeting?

Will it be on the 9th of May? If that's the case I won't be able to attend it, though @tiagormk (who's the author of these changes/feature) might be

ivanaamit commented:
Yes, it will be on May 9th at 4pm UTC. 

@tiagormk, would you be available to attend and discuss this improvement?

tiagormk commented:
> Yes, it will be on May 9th at 4pm UTC.
> 
> @tiagormk, would you be available to attend and discuss this improvement?

Unfortunately I have a conflict that time. But I could join the June meeting probably.

shinezyy commented:
Cool!
We are doing something similar: see gem5 issue #1164.

I notice that you have many empty implementations like 
```
Sequencer*
CHIGenericController::getCPUSequencer() const
{
    // CHIGenericController doesn't have a CPUSequencer
    return nullptr;
}
```

We are facing similar things. 
The root cause is that  `AbstractController` is the interface for a Ruby cache but not the interface for `an upstream` of Ruby cache. So we define a more basic class HasDownStream (has ruby cache downstream)  [here](https://github.com/shinezyy/gem5-1/tree/Extract-HasDownStream). This is only a draft.

I think there should be something defining a cleaner interface of `an upstream` of Ruby cache. I'd like to participate.

giactra commented:
> Cool! We are doing something similar: see gem5 issue #1164.
> 

Yes I saw the issue few days ago and I also noticed the similarity

> I notice that you have many empty implementations like
> 
> ```
> Sequencer*
> CHIGenericController::getCPUSequencer() const
> {
>     // CHIGenericController doesn't have a CPUSequencer
>     return nullptr;
> }
> ```
> 

Yes, that is mainly because this controller constitutes a pure virtual interface. We have a subclass we will upstream which overrides most of these methods to be able to transform a gem5 classic packet into a ruby packet. The reason why we don't commit to a sequencer in the CHIGenericController is because we want to be able to support the case of a sequencer-less configuration.

If I understood correctly this might be a useful scenario for you as well.

> We are facing similar things. The root cause is that `AbstractController` is the interface for a Ruby cache but not the interface for `an upstream` of Ruby cache. So we define a more basic class HasDownStream (has ruby cache downstream) [here](https://github.com/shinezyy/gem5-1/tree/Extract-HasDownStream). This is only a draft.
> 
> I think there should be something defining a cleaner interface of `an upstream` of Ruby cache. I'd like to participate.

Thanks I will look at your draft if I find some time. I will also keep you in the loop if there is an update. Most likely we will discuss this feature at the next gem5 developer meeting.

ivanaamit commented:
@Giactra, @Tiagormk: Are you available to present this at next week's developer meeting on Thursday, June 13th, at 4 PM UTC?

tiagormk commented:
> @giactra, @tiagormk: Are you available to present this at next week's developer meeting on Thursday, June 13th, at 4 PM UTC?

Yes, I can present this.

giactra committed:
mem-ruby: Add a CHI-TLM CacheController

This commit is extending the previously defined CHIGenericController
to implement a CacheController which acts as a bridge between the
AMBA TLM 2.0 implementation of CHI [1][2] with the gem5 (ruby) one.

In other words it translates AMBA CHI transactions into ruby
messages (which are then forwarded to the MessageQueues)
and viceversa.

ARM::CHI::Payload,         CHIRequestMsg
                     <-->  CHIDataMsg
ARM::CHI::Phase            CHIResponseMsg
                           CHIDataMsg

[1]: https://developer.arm.com/documentation/101459/latest
[2]: https://developer.arm.com/Architectures/AMBA#Downloads

Change-Id: I6f35e7b4ade4d0de1b5e5d2dbf73ce796a9f9fb6
Signed-off-by: Giacomo Travaglini <giacomo.travaglini@arm.com>

giactra committed:
mem-ruby: Add a CHI-TLM transaction Generator for testing

This commit is building over the CHI-TLM wrapping introduced
by the previous commit and it is adding a CHI traffic generator
as a SimObject.
This will get the python objects as input and it will forward
them to the TlmController to convert them into ruby CHI
messages

Change-Id: Ia67094c9bb880e37b24184313df546ecbaa3289f
Signed-off-by: Giacomo Travaglini <giacomo.travaglini@arm.com>

giactra committed:
mem-ruby: Wrap the CHI-TLM library with pybind11

This commit is wrapping the external AMBA CHI-TLM with pybind11
so that it will be possible to use its data structures/functions
from python.

More specifically we will be able to instantiate a ARM::CHI::Payload
and ARM::CHI::Phase from a gem5 config, with the end goal of being
able to configure a CHI transaction from python

Change-Id: I9587b445c21df44161fa3d9e09fc2651541b38bd
Signed-off-by: Giacomo Travaglini <giacomo.travaglini@arm.com>

giactra committed:
tests: Write unit-tests for ruby using the CHI-TLM library

This commit is adding two python files:

* ruby_mem_test.py is the canonical gem5 configuration script,
and it is an adaptation of the existing ruby_mem_test.py [1].
The main difference is the use of the TlmController as a
cache controller, and the use of TlmGenerator instead of
the MemTest memory tester. The config is minimally setting up
the system. The extent of the testing is specified in the second
python file:

* read_shared_unit.py: "unit-test" for the CHI ReadShared request
The file should be passed to the ruby_mem_test.py as cmdline
argument:

build/ARM/gem5.opt <>/ruby_mem_test.py <>/read_shared_unit.py

This is a simple testing file. We should ideally generate separate
test files for separate transactions/scenarios.
The test file can have whatever for inside, it only needs to comply
to the minimal interface required by the ruby_mem_test.py, which is
to define the following function:

def test_all(generator):

[1]: https://github.com/gem5/gem5/blob/stable/\
    configs/example/ruby_mem_test.py

Change-Id: I767ede9b8572f3eafe677c84da45fd904d77e319
Signed-off-by: Giacomo Travaglini <giacomo.travaglini@arm.com>

BobbyRBruce commented:
Is this waiting for review or are we still working on this?

giactra commented:
> Is this waiting for review or are we still working on this?

Hi Bobby,
We discussed this few developers meetings ago; the PR is actually ready for review, though @powerjg was asking to extend the PR with a use case for this CHIGenericController. We will post this anytime soon, but if you want to start reviewing this I am definitely not gonna stop you ;)

BobbyRBruce commented:
Next week (Wednesday) will be creating the Staging branch. Is this PR something you can complete soon and wish to get in?

I'm still a little unsure what needs done here but if I can help let just know

giactra commented:
> Next week (Wednesday) will be creating the Staging branch. Is this PR something you can complete soon and wish to get in?
> 
> I'm still a little unsure what needs done here but if I can help let just know

Thanks Bobby,

As I mentioned in my prior comment, I have posted an update to the PR with the requested usecase

> We discussed this few developers meetings ago; the PR is actually ready for review, though @powerjg was asking to extend the PR with a use case for this CHIGenericController. We will post this anytime soon, but if you want to start reviewing this I am definitely not gonna stop you ;)

As the usecase is quite significant (in terms of codesize) I had to modify the PR title and message to reflect the new scope of the change. I would have preferred to do this in a separate PR and get this merged long time ago but the upstream consensus was to have definition and usage in one big PR... I apologise for the confusion this might cause.

Anyway, the PR is now ready for review

tiagormk commented on src/mem/ruby/protocol/chi/tlm/controller.cc:
```diff
@@ -0,0 +1,424 @@
+    req_msg->m_txnId = phase.txn_id + (payload.lpid * 1024);
+    req_msg->m_ns = payload.ns;
+
+    sendRequestMsg(req_msg);
```
This (and the other `send*Msg` functions) currently assumes the underlying MessageBuffer is configured with unlimited capacity. I don't think this is a blocker for this PR, but it would be good to have a followup PR that adds some sort of link-level credit management to properly handle backpressure.

For instance, in CHIGenericController we could have a `virtual bool recvLinkCrd(unsigned crds, CHIChannel channel)` (or a separate one per channel type) that top-level can optionally implement to receive credits. It could be called from `CHIGenericController::init or initNetQueues` to return the initial number of credits (or 0 or UINT_MAX if the buffer is infinite), and every time a message is dequeued. We would need to register a callback with `MessageBuffer::registerDequeueCallback` to be informed of a dequeue.

I can create an issue to track this once it is merged.

tiagormk reviewed (approved):
Thanks @giactra !

LGTM (not sure it it counts as I've authored some of the commits). Also left a comment on better handling of backpressure, which is something that I recommend addressing in a future PR.

giactra commented:
Hi @powerjg @BobbyRBruce , can I have an update on whether there is a chance this will be merged in 24.1?

BobbyRBruce commented:
> Hi @powerjg @BobbyRBruce , can I have an update on whether there is a chance this will be merged in 24.1?

Yes! I've logged this as a PR we're going to let in directly to staging. I'm going to make a post about it soon, but essentially there are about 5 or 6 PRs we really want in but need work (or just reviews or testing), so while staging now exists (I've yet to make an announcement, but you can see it's a branch currently in the repo), we're going to permit some PRs (this one included)  to be merged into staging and therefore make it into the release.

Just seems like this release we have a lot of really nice features that just need just a _little more_ time, so we're going to give that time to PRs like this. WE'd really like to get most of these in over the next couple of weeks ideally though.

giactra commented:
> > Hi @powerjg @BobbyRBruce , can I have an update on whether there is a chance this will be merged in 24.1?
> 
> Yes! I've logged this as a PR we're going to let in directly to staging. I'm going to make a post about it soon, but essentially there are about 5 or 6 PRs we really want in but need work (or just reviews or testing), so while staging now exists (I've yet to make an announcement, but you can see it's a branch currently in the repo), we're going to permit some PRs (this one included) to be merged into staging and therefore make it into the release.
> 
> Just seems like this release we have a lot of really nice features that just need just a _little more_ time, so we're going to give that time to PRs like this. WE'd really like to get most of these in over the next couple of weeks ideally though.

Thanks Bobby, makes sense. Let me know if I can help somehow if there's anything that needs reviewing

giactra commented:
Hi @BobbyRBruce and @powerjg please have a look at the new iteration of the PR.
I have basically added some documentation on the Controller class and I stopped using Ruby.create_system as requested.

Please note: I am still using Ruby.setup_memory_controller. Let me know if you want me to remove that usage as well
